README.md
# animated_battery_gauge widget

Very simple to use animated battery gauge widget.

## Screenshots
- Horizontal gauge mode
![](https://github.com/shiniwat/animated_battery_gauge/blob/main/demo/animated_battery_horizontal.gif)
- Vertical gauge mode
![](https://github.com/shiniwat/animated_battery_gauge/blob/main/demo/animated_battery_vertical.gif)
- Horizontal grid mode
![](https://github.com/shiniwat/animated_battery_gauge/blob/main/demo/animated_grid_mode_horizontal.gif)

## Features

- Both animated and static widgets exist. AnimatedBatteryGauge is animated version, and BatteryGauge is non-animated version.
- Both horizontal and vertical widgets are supported. See screenshot above.
- Starting from version 0.2.0, the widget takes mode parameter, which takes BatteryGaugePaintMode. You can take either gauge, grid, or none as the mode parameter.

## Getting started

Please see example app under example folder.

## Usage

It is as simple as using BatteryGauge or AnimatedBatteryGauge as your child widget.

```dart
Widget build(BuildContext context) {
  return CupertinoPageScaffold(
    navigationBar: CupertinoNavigationBar(
      middle: Text('Animated Battery Demo'),
    ),
    child: Center(
      child: AnimatedBatteryGauge(
        duration: Duration(seconds: 2),
        value: 60,
        size: Size(150, 70),
        borderColor: CupertinoColors.systemGrey,
        valueColor: CupertinoColors.activeGreen,
        mode: BatteryGaugePaintMode.gauge,
        hasText: true,
      ),
    ),
  );
}
```


CHANGELOG.md
## 0.2.0
- Add BatteryGaugePaintMode as the parameter.
- Add grid mode for painting the battery value.

## 0.1.0
- Initial release.



test/animated_battery_gauge_test.dart
import 'package:animated_battery_gauge/battery_gauge.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/cupertino.dart';

void main() {
  const batteryGauge = BatteryGauge(
      size: Size(100, 60),
      value: 50,
      borderColor: CupertinoColors.systemGrey,
      valueColor: CupertinoColors.activeBlue,
      hasText: true);
  const testWidget = CupertinoApp(home: Center(child: batteryGauge));
  testWidgets('AnimatedBatteryGauge test ', (widgetTester) async {
    await widgetTester.pumpWidget(testWidget);
    final valueFinder = find.text('50%');
    expect(valueFinder, findsOneWidget);
  });
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';

import 'battery_gauge_demo.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      title: 'Flutter Demo',
      theme: CupertinoThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primaryColor: CupertinoColors.activeBlue,
      ),
      home: StartWidget(), //MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class StartWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(
        middle: Text('Animated Battery Demo'),
      ),
      child: Center(
        child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              CupertinoButton(
                onPressed: () {
                  Navigator.of(context)
                      .push(CupertinoPageRoute(builder: (context) {
                    return BatteryGaugeDemo(
                      isHorizontal: true,
                      isGrid: false,
                    );
                  }));
                },
                child: Text('Show Battery Horizontal w/ gauge'),
              ),
              CupertinoButton(
                onPressed: () {
                  Navigator.of(context)
                      .push(CupertinoPageRoute(builder: (context) {
                    return BatteryGaugeDemo(
                      isHorizontal: false,
                      isGrid: false,
                    );
                  }));
                },
                child: Text('Show Battery Vertical w/ gauge'),
              ),
              CupertinoButton(
                onPressed: () {
                  Navigator.of(context)
                      .push(CupertinoPageRoute(builder: (context) {
                    return BatteryGaugeDemo(
                      isHorizontal: true,
                      isGrid: true,
                    );
                  }));
                },
                child: Text('Show Battery Horizontal w/ grid mode'),
              ),
              CupertinoButton(
                onPressed: () {
                  Navigator.of(context)
                      .push(CupertinoPageRoute(builder: (context) {
                    return BatteryGaugeDemo(
                      isHorizontal: false,
                      isGrid: true,
                    );
                  }));
                },
                child: Text('Show Battery Vertical w/ grid mode'),
              ),
            ]),
      ),
    );
  }
}


lib/battery_gauge.dart
library animated_battery_gauge;

import 'package:flutter/cupertino.dart';

enum BatteryGaugePaintMode {
  gauge,
  grid,
  none,
}

class BatteryGauge extends StatefulWidget {
  final Size size;
  final int value;
  final Color borderColor;
  final Color valueColor;
  final BatteryGaugePaintMode mode;
  final bool hasText;
  final TextStyle? textStyle;
  final bool drawBarForExtraValue;
  const BatteryGauge(
      {Key? key,
      required this.size,
      required this.value,
      required this.borderColor,
      required this.valueColor,
      this.mode = BatteryGaugePaintMode.gauge,
      this.hasText = false,
      this.textStyle,
      this.drawBarForExtraValue = false})
      : super(key: key);

  @override
  createState() => _BatteryGaugeState();
}

class _BatteryGaugeState extends State<BatteryGauge> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    final margin = widget.size.width * 0.04 * 2;
    final isHorizontal = widget.size.width >= widget.size.height;
    return SizedBox(
      width: (isHorizontal) ? widget.size.width + margin : widget.size.width,
      height: (isHorizontal) ? widget.size.height : widget.size.height + margin,
      child: Stack(
        children: [
          CustomPaint(
            size: widget.size,
            painter: _BatteryGaugePainter(
                value: widget.value,
                borderColor: widget.borderColor,
                valueColor: widget.valueColor,
                mode: widget.mode,
                drawBarForExtraValue: widget.drawBarForExtraValue),
          ),
          if (widget.hasText)
            Positioned(
              child: Center(
                child: Text('${widget.value.toString()}%',
                    style: widget.textStyle ??
                        CupertinoTheme.of(context).textTheme.textStyle),
              ),
            ),
        ],
      ),
    );
  }
}

class _BatteryGaugePainter extends CustomPainter {
  final int value;
  final Color borderColor;
  final Color valueColor;
  final BatteryGaugePaintMode mode;
  final bool drawBarForExtraValue;

  const _BatteryGaugePainter(
      {required this.value,
      required this.borderColor,
      required this.valueColor,
      required this.mode,
      required this.drawBarForExtraValue})
      : super();

  @override
  void paint(Canvas canvas, Size size) {
    double stroke = size.width * 0.04;
    final isHorizontal = size.width >= size.height;

    var paintBorder = Paint()
      ..color = borderColor
      ..shader
      ..strokeWidth = stroke
      ..style = PaintingStyle.stroke;
    var paintValue = Paint()
      ..color = valueColor
      ..shader
      ..style = PaintingStyle.fill;

    final margin = stroke * 2;
    // render battery rect and connector
    if (isHorizontal) {
      // render battery rect
      canvas.drawRRect(
        RRect.fromRectAndRadius(Rect.fromLTWH(0, 0, size.width, size.height),
            Radius.circular(size.width * 0.05)),
        paintBorder,
      );

      /// plus connector
      canvas.drawRRect(
        RRect.fromRectAndRadius(
            Rect.fromLTWH(size.width, size.height * 0.333, stroke * 2,
                size.height * 0.333),
            Radius.circular(size.width * 0.02)),
        paintBorder,
      );
    } else {
      // vertical
      // render battery rect
      canvas.drawRRect(
        RRect.fromRectAndRadius(
            Rect.fromLTWH(0, margin, size.width, size.height),
            Radius.circular(size.width * 0.05)),
        paintBorder,
      );

      /// plus connector
      canvas.drawRRect(
        RRect.fromRectAndRadius(
            Rect.fromLTWH(
                size.width * 0.333, 0, size.width * 0.333, stroke * 2),
            Radius.circular(size.width * 0.02)),
        paintBorder,
      );
    }

    // render value
    switch (mode) {
      case BatteryGaugePaintMode.none:
      case BatteryGaugePaintMode.gauge:
        if (isHorizontal) {
          // render value
          canvas.drawRRect(
            RRect.fromRectAndRadius(
                Rect.fromLTWH(0, 0, size.width * value / 100, size.height),
                Radius.circular(size.width * 0.05)),
            paintValue,
          );
        } else {
          // render value
          final margin = stroke * 2;
          canvas.drawRRect(
            RRect.fromRectAndRadius(
                Rect.fromLTWH(
                    0,
                    margin + size.height - size.height * value / 100,
                    size.width,
                    size.height * value / 100),
                Radius.circular(size.width * 0.05)),
            paintValue,
          );
        }
        break;

      case BatteryGaugePaintMode.grid:
        var paintInactive = Paint()
          ..color = CupertinoColors.systemGrey5
          ..shader
          ..style = PaintingStyle.fill;
        var numBar = value ~/ 10;
        if (drawBarForExtraValue && value.toInt() % 10 > 0) {
          numBar++;
        }
        final list = _getRectList(size, value, stroke);
        for (int i = 0; i < list.length; i++) {
          canvas.drawRect(list[i], (i < numBar) ? paintValue : paintInactive);
        }
        break;
    }

    if (mode == BatteryGaugePaintMode.gauge) {
      // draw gauge
      var paintGauge = Paint()
        ..color = CupertinoColors.systemGrey2
        ..shader
        ..strokeWidth = 1
        ..style = PaintingStyle.stroke;
      if (isHorizontal) {
        for (var i = 1; i < 5; i++) {
          var offset1 = Offset(size.width / 5 * i, size.height / 10);
          var offset2 = Offset(size.width / 5 * i, size.height / 10 * 9);
          canvas.drawLine(offset1, offset2, paintGauge);
        }
      } else {
        final margin = stroke * 2;
        for (var i = 1; i < 5; i++) {
          var offset1 = Offset(size.width / 10, margin + size.height / 5 * i);
          var offset2 =
              Offset(size.width / 10 * 9, margin + size.height / 5 * i);
          canvas.drawLine(offset1, offset2, paintGauge);
        }
      }
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }

  List<Rect> _getRectList(Size size, int value, double stroke) {
    var list = <Rect>[];
    const margin = 2.0;
    if (size.width >= size.height) {
      final barWidth = (size.width - margin * 11 - stroke * 2) / 10;
      for (int i = 0; i < 10; i++) {
        final offset = stroke + (i + 1) * margin + i * barWidth;
        final rect = Rect.fromLTRB(offset, stroke + margin, offset + barWidth,
            size.height - stroke - margin);
        list.add(rect);
      }
    } else {
      final connectorHeight = stroke * 2;
      final barHeight = (size.height - margin * 11 - stroke * 2) / 10;
      for (int i = 0; i < 10; i++) {
        final bottomOffset = size.height -
            connectorHeight -
            stroke -
            (i + 1) * margin -
            (i - 1) * barHeight;
        final rect = Rect.fromLTRB(stroke + margin, bottomOffset - barHeight,
            size.width - stroke - margin, bottomOffset);
        list.add(rect);
      }
    }
    return list;
  }
}


lib/animated_battery_gauge.dart
library animated_battery_gauge;

import 'package:flutter/cupertino.dart';
import 'package:animated_battery_gauge/battery_gauge.dart';

class AnimatedBatteryGauge extends ImplicitlyAnimatedWidget {
  final double value;
  final Size size;
  final Color borderColor;
  final Color valueColor;
  final BatteryGaugePaintMode mode;
  final bool hasText;
  final TextStyle? textStyle;
  final bool drawBarForExtraValue;

  const AnimatedBatteryGauge(
      {Key? key,
      required super.duration,
      required this.value,
      required this.size,
      required this.borderColor,
      required this.valueColor,
      this.mode = BatteryGaugePaintMode.gauge,
      this.hasText = false,
      this.textStyle,
      this.drawBarForExtraValue = false})
      : super(key: key);

  @override
  createState() => _AnimatedBatteryGaugeState();
}

class _AnimatedBatteryGaugeState
    extends AnimatedWidgetBaseState<AnimatedBatteryGauge> {
  Tween<double>? _valueTween;

  @override
  void initState() {
    super.initState();
    controller
      ..value = 0.0
      ..forward().whenCompleteOrCancel(() {
        // do nothing so far.
      });
  }

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      child: Builder(builder: (context) {
        final drawValue = _valueTween!.evaluate(animation).clamp(
              0,
              100,
            );
        return BatteryGauge(
          size: widget.size,
          value: drawValue.toInt(),
          borderColor: widget.borderColor,
          valueColor: widget.valueColor,
          mode: widget.mode,
          hasText: widget.hasText,
          textStyle: widget.textStyle,
          drawBarForExtraValue: widget.drawBarForExtraValue,
        );
      }),
    );
  }

  @override
  void forEachTween(TweenVisitor<dynamic> visitor) {
    _valueTween = visitor(
      _valueTween,
      widget.value,
      (dynamic value) => Tween<double>(
        begin: 0.0,
        end: widget.value,
      ),
    ) as Tween<double>;
  }
}


example/lib/battery_gauge_demo.dart
import 'package:animated_battery_gauge/battery_gauge.dart';
import 'package:flutter/cupertino.dart';
import 'package:animated_battery_gauge/animated_battery_gauge.dart';

class BatteryGaugeDemo extends StatelessWidget {
  final bool isHorizontal;
  final bool isGrid;
  const BatteryGaugeDemo(
      {Key? key, required this.isHorizontal, required this.isGrid})
      : super(key: key);
  @override
  Widget build(BuildContext context) {
    if (isGrid) {
      return CupertinoPageScaffold(
          navigationBar: CupertinoNavigationBar(
            middle: Text('Grid style Battery Gauge'),
          ),
          child: Center(
            child: AnimatedBatteryGauge(
              duration: Duration(seconds: 2),
              value: (isHorizontal) ? 60 : 42,
              size: (isHorizontal) ? Size(150, 70) : Size(70, 150),
              borderColor: CupertinoColors.systemGrey,
              valueColor: (isHorizontal)
                  ? CupertinoColors.activeGreen
                  : CupertinoColors.activeBlue,
              hasText: true,
              mode: BatteryGaugePaintMode.grid,
            ),
          ));
    } else {
      return CupertinoPageScaffold(
        navigationBar: CupertinoNavigationBar(
          middle: Text('Animated Battery Demo'),
        ),
        child: Center(
          child: AnimatedBatteryGauge(
            duration: Duration(seconds: 2),
            value: (isHorizontal) ? 60 : 42,
            size: (isHorizontal) ? Size(150, 70) : Size(70, 150),
            borderColor: CupertinoColors.systemGrey,
            valueColor: (isHorizontal)
                ? CupertinoColors.activeGreen
                : CupertinoColors.activeBlue,
            hasText: true,
          ),
        ),
      );
    }
  }
}
